auth_mellon_cache.c

```diff
@@ -758,6 +901,29 @@ void am_cache_delete(request_rec *r, am_cache_entry_t *cache)
     am_cache_unlock(r, cache);
 }
 
+/* This function deletes the specified keys from the session store
+ * and releases the locks.
+ * Equivalent to performing am_cache_delete on multiple sessions.
+ *
+ * Parameters:
+ *  request_rec *r            The request we are processing.
+ *  apr_array_header_t *sessions   The entry we are deleting.
+ *
+ * Returns:
+ *  Nothing.
+ */
+void am_cache_delete_sessions(request_rec *r, apr_array_header_t *sessions)
+{
+    am_cache_entry_t *cache;
+    int i;
+    for (i = 0; i < sessions->nelts; i++) {
+        cache = APR_ARRAY_IDX(sessions, i, am_cache_entry_t *);
+        am_diag_log_cache_entry(r, 0, cache, "delete session");
+        cache->key[0] = '\0';
+        cache->access = apr_time_now();
+    }
+    am_cache_mutex_unlock(r);
```

Comment:
this is bad practice ... hiding a lock this deep (two nested) into a utility function is a sure way to get bugs later as it will be very difficult to remember and figure out.

As a general rule locking and unlocking should happen in the same calling function, so that it is very explicit what is the context of the locked code, and all error conditions can be properly checked.

Please rework this code to get the unlock out of this call and up a few levels.

Reply:
I agree. However there is a reason why I decided on this code.
The code before modification was to lock with am_get_request_session_nameid() and then unlock with am_delete_request_session(). 
I tried to make it consistent with the existing code.

I couldn't come up with a good fix, so I need someone else to make a better fix.